src/machine/machined-core.c

```diff
@@ -175,3 +181,220 @@ void manager_enqueue_gc(Manager *m) {
+                        if ((size_t) n != sizeof(family) + FAMILY_ADDRESS_SIZE(family))
+                                return -EIO;
+
+                        r = add_local_address(&list,
+                                              &n_list,
+                                              /* ifindex = */ 0,
+                                              /* scope = */ '\0',
+                                              family,
+                                              &in_addr);
+                        if (r < 0)
+                                return r;
+                }
+
+                r = wait_for_terminate_and_check("(sd-addrns)", child, 0);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to wait for child: %m");
+                if (r != EXIT_SUCCESS) {
+                        log_debug("Child died abnormally, exit code=%d", r);
+                        return -ESHUTDOWN;
+                }
+
+                *ret_addresses = TAKE_PTR(list);
+                return (int) n_list;
+        }
+
+        default:
+                return -EOPNOTSUPP;
+        }
+}
+
+#define EXIT_NOT_FOUND 2
+
+int machine_get_os_release(Machine *machine, char ***ret_os_release) {
+        _cleanup_strv_free_ char **l = NULL;
+        int r;
+
+        assert(machine);
+        assert(ret_os_release);
+
+        switch (machine->class) {
+
+        case MACHINE_HOST:
+                r = load_os_release_pairs(NULL, &l);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to load OS release information: %m");
+
+                break;
+
+        case MACHINE_CONTAINER: {
+                _cleanup_close_ int mntns_fd = -EBADF, root_fd = -EBADF, pidns_fd = -EBADF;
+                _cleanup_close_pair_ int pair[2] = EBADF_PAIR;
+                _cleanup_fclose_ FILE *f = NULL;
+                pid_t child;
+
+                r = pidref_namespace_open(&machine->leader,
+                                          &pidns_fd,
+                                          &mntns_fd,
+                                          /* ret_netns_fd = */ NULL,
+                                          /* ret_userns_fd = */ NULL,
+                                          &root_fd);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to open namespace: %m");
+
+                if (socketpair(AF_UNIX, SOCK_SEQPACKET, 0, pair) < 0)
+                        return log_debug_errno(-errno, "Failed to call socketpair(): %m");
+
+                r = namespace_fork("(sd-osrelns)", "(sd-osrel)", NULL, 0, FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGKILL,
+                                   pidns_fd, mntns_fd, -1, -1, root_fd,
+                                   &child);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to fork(): %m");
+                if (r == 0) {
+                        int fd = -EBADF;
```

Comment:
Missing `_cleanup_close_` here. It technically doesn't matter but it's good to be consistent